### PR TITLE
core: improve updating trees and preparing assets

### PR DESCRIPTION
**Pull request by KeefeL**
### Description

1. This PR introduces intermediating state root when committing a new block, and deletes the update tree after each transaction has been applied.
2. This PR makes the preparation for the related accounts and assets precise, with no redundant initialization.

### Rationale

1. In the current implementation, tx executors will update state trees after each tx has been applied. Actually, we need to update the state tree only when we commit a block, there is no need to compute the state root of each transaction in the committer process. This PR can avoid lots of unnecessary computing of state root, and improve the efficiency of the committer.
2. When we prepare the related accounts and assets, the assets of some accounts may doesn't affected, so we should delete the unnecessary initialization.

### Example

NA

### Changes

Notable changes:
* Delete update tree for each tx executor
* Make prepare precise

### CODE REVIEW

File: core/statedb/state_cache.go
Comment (yutianwu) [inline core/statedb/state_cache.go:76]: you can change these two lines into one

File: core/statedb/statedb.go
Comment (yutianwu) [inline core/statedb/statedb.go:387]: assetId is the index of the slice, not the value

File: core/executor/base_executor.go
Comment (yutianwu) [inline core/executor/base_executor.go:8]: fix the imports

File: core/statedb/statedb.go
Comment (yutianwu) [inline core/statedb/statedb.go:542]: I'd like `updateNft` rather than`nftIntermediateRoot`

File: core/block_processor.go
Comment (guagualvcha) [inline core/block_processor.go:54]: For committer, there is no need to `UpdateTrees`, but it is a necessary operation for witness. Because we are going to reuse the code, so I think we need think about how witness is going to reuse  the code before merging this PR.

File: core/executor/base_executor.go
Comment (guagualvcha) [inline core/executor/base_executor.go:43]: I will suggest return the err rather than `internal error`

File: core/executor/base_executor.go
Comment (guagualvcha) [inline core/executor/base_executor.go:123]: Looks like every exeutor will all `core/executor/base_executor.go`, I think we can simply it.

File: core/block_processor.go
Comment (KeefeL) [reply core/block_processor.go:54]: For witness processor, can call IntermediateRoot after each transaction to get the state root, and then clear the dirty objects to avoid redundant computation.

File: core/executor/base_executor.go
Comment (KeefeL) [reply core/executor/base_executor.go:43]: This is reused by api server, need to format the error return to the user.

File: core/statedb/state_cache.go
Comment (KeefeL) [reply core/statedb/state_cache.go]: fixed

File: core/executor/base_executor.go
Comment (KeefeL) [reply core/executor/base_executor.go]: fixed

File: core/statedb/statedb.go
Comment (KeefeL) [reply core/statedb/statedb.go]: fixed

File: core/block_processor.go
Comment (KeefeL) [reply core/block_processor.go:54]: Add a flag for IntermediateRoot, then we can reuse it.

File: core/executor/base_executor.go
Comment (KeefeL) [reply core/executor/base_executor.go:125]: fixed

File: core/statedb/state_cache.go
Comment (guagualvcha) [inline core/statedb/state_cache.go:36]: Please use lower case.

**Review by yutianwu — APPROVED** [review]

File: core/statedb/statedb.go
Comment (guagualvcha) [inline core/statedb/statedb.go:440]: Why need `cleanDirty`? seems we only use it as false.

File: core/statedb/statedb.go
Comment (guagualvcha) [inline core/statedb/statedb.go:450]: For the failed tx, it will also mark the dirty map, will it cause problem?

File: core/statedb/statedb.go
Comment (KeefeL) [reply core/statedb/statedb.go:440]: When we want to compute a state root after a transaction, we can set cleanDirty to true. Because stateCache itself is block level.

File: core/statedb/statedb.go
Comment (KeefeL) [reply core/statedb/statedb.go:450]: It won't, the failed tx only update the executor's dirty map, it won't sync to the state cache's dirty map.

File: core/statedb/state_cache.go
Comment (KeefeL) [reply core/statedb/state_cache.go:36]: I will fix this in the next refactor tx PR, after this PR merged, I will gonna to rebase this, and fix this one.

**Review by guagualvcha — APPROVED** [review]

**Pull request merged**
